File: j1077-recursive.py

#  j1077-recursive.py
#
#  Created by Yongduek Seo on 2016. 4. 25..
#  Copyright Yongduek Seo. All rights reserved.
#  yndk@sogang.ac.kr

import sys

class jewelry:
    def __init__(self, w, v):
        self.weight=w
        self.value=v
    #end def
#end class definition

N,W = map(int, sys.stdin.readline().split())
packTable = [-1 for i in range (W+1)]

#print (N)
#print (W)
#print (packTable)

jewList = []

for i in range(N) :
    w,v = map(int, sys.stdin.readline().split())
    jewList.append(jewelry(w,v))

#for i in range(N) :
#    print (jewList[i].weight, jewList[i].value)

def pack(w) :
    maxval = 0
    if packTable[w] < 0 :
        for j in jewList :
            weight_diff = w - j.weight
            if weight_diff >= 0 :
                total_val = j.value + pack(weight_diff)
                if maxval < total_val :
                    maxval = total_val
                #endif
            #endif
        #end-for
        packTable[w] = maxval;
    #endif
    return packTable[w]
# end def pack(w)

maxval = pack(W)
print (maxval)

# eof #

File: j1077-recursive.cpp
//
//  j1077-recursive.cpp
//  UVa-XCode
//
//  Created by Yongduek Seo on 2016. 4. 25..
//  Copyright © 2016년 Yongduek Seo. All rights reserved.
//

#include <stdio.h>
#include <iostream>
#include <vector>
using namespace std;

struct jewelry {
    int weight, value;
};

vector<jewelry> jews;

const int maxWeight = 10000;
int packTable[maxWeight+1];

int pack(int w)
{
//    if (w<=0) return 0;
    
    int maxval = 0;
    if (packTable[w]<0)
        {
        for (int i=0; i<jews.size(); i++)
            {
            int weight_possible = w - jews[i].weight;
            if (weight_possible < 0) continue;
            int total_val = jews[i].value + pack (weight_possible);
            if (maxval < total_val) maxval = total_val;
            }
        packTable[w] = maxval;
        }
    
    maxval = packTable[w];
    return maxval;
}
int main()
{
    int N, W;
    cin >> N >> W;
    
    for (int i=0; i<=W; i++) packTable[i] = -1;
    
    for (int i=0; i<N; i++) {
        int weight, val;
        cin >> weight >> val;
        jews.push_back(jewelry {weight, val});
    }
    
    int maxval = pack (W);
    
    cout << maxval << endl;

//    cout << "--" << endl;
//    for (int i=0; i<=W; i++)
//        cout << i << ": " << packTable[i] << endl;

    return 0;
}
// eof //